[Neo Weihong] Duke Increments

whneo97 opened the pull request:
Contains all updates to my Duke repository from Levels 1 to 9, and intermediate increments.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

whneo97 committed:
Gradle

whneo97 committed:
Merge branch 'gradle' of https://github.com/whneo97/duke

# Conflicts:
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew.bat

whneo97 committed:
Added checkstyle to gradle build file

whneo97 committed:
plugins: add shadow plugin

create test files and run text UI test

whneo97 committed:
Gradle build file: Allow gradle to handle scanner

whneo97 committed:
Edit Hello Message

whneo97 committed:
LEVEL 1: GREET

whneo97 committed:
LEVEL 1: ECHO, EXIT

whneo97 committed:
LEVEL 2: ADD, LIST

whneo97 committed:
LEVEL 3: DONE (primitive)

whneo97 committed:
LEVEL 3: DONE (using classes)

whneo97 committed:
LEVEL 4: ToDos, Events, Deadlines (without Inheritance)

whneo97 committed:
LEVEL 4: ToDos, Events, Deadlines (using inheritance)

whneo97 committed:
LEVEL 4: ToDos, Events, Deadlines (using inheritance); corrected and additional minor changes to toString() methods

whneo97 committed:
LEVEL 5: EXCEPTIONS

whneo97 committed:
LEVEL 5: EXCEPTIONS

whneo97 committed:
LEVEL 5: EXCEPTIONS (additional exceptions particularly for "done" command)

Number entered after done is ensured to be within bounds of indexes in the list (where index is numbered from 1)

whneo97 committed:
LEVEL 6: DELETE

Relocation of exceptions to individual classes

whneo97 committed:
Included javadoc comments for all classes.

whneo97 committed:
Set constructors and methods to public. Included javadocs

whneo97 committed:
Level 7 Save
And minor bug fixes

whneo97 committed:
Level 8 Dates and Times

whneo97 committed:
Level 7 Load

whneo97 committed:
Merge branch 'branch-Level-8' into branch-Level-7

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/Task.java

whneo97 committed:
Merge branch 'branch-Level-7'

whneo97 committed:
Level 8.1 minor bug fixes

whneo97 committed:
A-MoreOOP.
Major restructuring of classes.

whneo97 committed:
A-Enums

whneo97 committed:
A-Packages

whneo97 committed:
A-JUnit.
Minor Bug Fixes.

whneo97 committed:
A-Jar

whneo97 committed:
A-JavaDoc

whneo97 committed:
A-CodingStandard

whneo97 committed:
Made edits to javadoc and checkstyle

whneo97 committed:
Restructure packages and modify javadoc

whneo97 committed:
A-CodingStandard

whneo97 committed:
A-CodingStandard for Tests

whneo97 committed:
Level-9

whneo97 committed:
Modify Level-9 Find

whneo97 committed:
Merge branch 'branch-Level-9' into branch-A-CodingStandard

# Conflicts:
#	src/main/java/duke/dateandtime/DateAndTime.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/ui/Ui.java

whneo97 committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9 (recommit)

whneo97 committed:
Resolve conflicts in merge between Level-9 and A-Coding-Style branch.

whneo97 committed:
Resolve further conflicts between Level-9 and A-Coding-Style

whneo97 committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9 and resolve conflicts.

whneo97 committed:
Update duke jar file

whneo97 committed:
Merge branch 'branch-Level-9' to master

jyx11011 commented on src/main/java/seedu/duke/validation/Validation.java:
```diff
@@ -0,0 +1,203 @@
+            isLeapYear = true;
+        } else if (year % 100 == 0 && year % 400 == 0) {
+            isLeapYear = true;
+        }
```
It seems that a year which is divisible by 100 but not divisible by 400 are also determined as leap year here.

whneo97 replied:
Thanks for helping to spot the logic bug!

Have corrected the code to behave as follows instead:

```suggestion
        if (year % 4 == 0 && year % 100 != 0) {
            isLeapYear = true;
        } else if (year % 400 == 0) {
            isLeapYear = true;
        }
```

jyx11011 commented on src/main/java/seedu/duke/validation/Validation.java:
```diff
@@ -0,0 +1,203 @@
+                        + year + ".");
+            }
+        }
+
```
Very detailed and accurate information shown to user!

whneo97 replied:
Thanks for the positive feedback! Glad to know it's helpful to the user :)

jyx11011 commented on src/main/java/seedu/duke/task/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+                searchResults.add(task);
+            }
+        }
+        return searchResults;
```
Could you explain more about the logic for searching a string please?

whneo97 replied:
Alright sure, have added them as javadoc comments and regular comments for the most recent commit!

```suggestion
            /**
     * Returns a TaskList with search results for this TaskList.
     * Search keywords may contain the task description of tasks to be searched for,
     * the status of whether tasks have been marked as done, the types of task, date / time
     * (in DD/MM/YYYY and HHMM format respectively), and separator keywords for the type of task
     * (eg. "at" for event and "by" for deadline).
     * If none of the task descriptions, types, done statuses, dates/ times etc. of all the tasks match the
     * search keyword, the method returns an empty TaskList.
     * @param taskString Keywords to be used to search this TaskList for relevant Tasks.
     * @return TaskList containing keywords searched for in the given taskString.
     */
    public TaskList find(String taskString) {
        TaskList searchResults = new TaskList();
        for (int i = 0; i < taskList.size(); i++) {
            Task task = taskList.get(i);

            // If search keywords contain the task description or vice versa.
            // the task is added to the list of search results.
            if (task.getTaskString().contains(taskString)
                    || taskString.contains(task.getTaskString())) {
                searchResults.add(task);
            // If search keyword is "done" (i.e. search for tasks which have been marked as completed,
            // the task is added to the list of search results if it has been marked as done.
            } else if (task.getIsDone() == true && taskString.equals("done")) {
                searchResults.add(task);
            // If search keyword is "undone" or "not done"
            // (i.e. search for tasks which have yet to be marked as completed),
            // the task is added to the list of search results if it has yet to be marked as done.
            } else if (task.getIsDone() == false && (taskString.equals("undone")
                    || taskString.contains("not done"))) {
                searchResults.add(task);
            // If search keywords contain type of task (eg. deadline, event etc.),
            // the task is added to the list of search results if it's type matches the type to be searched for.
            // Type of task can be in any case (uppercase / lowercase).
            } else if (task.getType().toString().toLowerCase().contains(taskString.toLowerCase())
                    || taskString.toLowerCase().contains(task.getType().toString().toLowerCase())) {
                searchResults.add(task);
            // If search keywords are contained in the date (DD/MM/YYYY format) or time (HHMM format) attributes
            // of the task, the task is added to the list of search results.
            } else if (task.getDateAndTime() != null
                    && task.getDateAndTime().toString().contains(taskString)
                    || taskString.contains(task.getTaskString())) {
                searchResults.add(task);
            // If search keyword is "at" for an event-type task or search keyword is "by" for an event-type task,
            // the task is added to the list of search results if it is an event-type or
            // deadline type task respectively.
            } else if ((task.getType().toString().equals("EVENT") && taskString.equals("at"))
                    || (task.getType().toString().equals("DEADLINE") && taskString.equals("by"))) {
                searchResults.add(task);
            }
        }
        return searchResults;
    }
```

whneo97 committed:
Implement simple GUI as according to tutorial (without having implemented actual functionalities of Duke)

whneo97 committed:
Fix bugs for Tutorial 4 of JavaFx Tutorials

whneo97 committed:
Add functional GUI support for all existing duke operations, with relocation of main method to Launcher

whneo97 committed:
Update font and add resizing features.

whneo97 committed:
Add user and duke icons, in addition to enlarging font size.

whneo97 committed:
Add java background picture and aesthetics for scrollbar using css.

whneo97 committed:
Update javadoc and run checkstyle.

whneo97 committed:
Add 'help' command for user to see full list of available commands, and implement VarArgs in the process.

whneo97 committed:
Create notification message for unsuccessful location and restoration of data file.

whneo97 committed:
Update help descriptions (minor changes).

whneo97 committed:
Standardise code in css file.

whneo97 committed:
Include help description for 'bye' command, and correct grammar mistake for help description of 'list' command.

whneo97 committed:
Update Duke jar file.

whneo97 committed:
Add colour and set font for Duke logo for alignment, and include appropriate padding for system's output response.

whneo97 committed:
Experiment with colours in the dialog between user and Duke.

whneo97 committed:
Merge branch 'master' into branch-Font-Colour

whneo97 committed:
Merge branch 'branch-Font-Colour' into branch-Varargs

whneo97 committed:
Merge branch 'branch-Varargs' into branch-CheckStyle

whneo97 committed:
Merge branch 'branch-CheckStyle' into branch-Level-10

whneo97 committed:
Update comments for colour experimentation, code indentation using checkstyle and jar file.

whneo97 committed:
Amend typographical error description in help description for 'event'.

whneo97 committed:
Reorganised classes into packages following the directory structure of AddressBook3.

whneo97 committed:
Fix bugs in determination of leap years. Add comments to explain logic for handling search keywords.

whneo97 committed:
Add "Here are the tasks in your list:" line in Ui that was previously accidentally deleted.

whneo97 committed:
Remove javafx libraries and Photoshop file for GUI background image from duke project directory.

whneo97 committed:
Update background photo credit description in css file.

whneo97 committed:
Add assertions.
Implement abstraction for splitting of tokens in Deadline and Event.
Remove redundant variables in Add command.
Reclassify DukeExceptions into directories of different DukeException types.

whneo97 committed:
Correct code violations using checkstyle.

whneo97 committed:
Merge branch 'master' into A-CodeQuality

whneo97 committed:
Merge branch 'branch-A-CodeQuality' into branch-A-Assertions

whneo97 committed:
Add support for deleting multiple tasks and marking multiple tasks as done and undone.
Update help catalogue to include new commands and functions.

whneo97 committed:
Merge branch 'branch-A-Assertions'

whneo97 committed:
Add support to mark all tasks as done, undone or delete all.

whneo97 committed:
Implement draft version of Undo and Redo commands

whneo97 committed:
Fix undo and redo bugs regarding done and undone status.

whneo97 committed:
Fix minor bugs for deleting all tasks.

whneo97 committed:
Add support for sorting by various criteria and checking for duplicates.
Add javadoc for TaskListHistory (that stores tasks for undo/ redo operations).

whneo97 committed:
Refine help guide to include examples for user.
Fix typo errors in help guide.

whneo97 committed:
Add travis yml file

whneo97 committed:
Update travis yml file.

whneo97 committed:
Test travis (using deliberate styling error).

whneo97 committed:
Confirm travis ci works and correct deliberate styling error.

whneo97 committed:
Add java dependencies in gradle build file.

whneo97 committed:
Fix Storage issues for cross-os compatability with both Mac and Windows.

whneo97 committed:
Add 'final' keyword to a constant variable.

whneo97 committed:
Set preferred size of Main Window to be a golden rectangle.

whneo97 committed:
Merge branch 'master' of https://github.com/whneo97/duke

whneo97 committed:
Fix checkstyle violation for naming constants.

whneo97 committed:
Fix bugs in output list of Done and Undone Command.
Correct typo errors in javadoc of TaskList.

whneo97 committed:
Add random generator command to generate random list of tasks as a sample for experimental purposes.

whneo97 committed:
Fix checkstyle violations.

whneo97 committed:
Merge pull request #6 from whneo97/branch-random-tasklist-generator

Add random generator command to generate random list of tasks.
Fix checkstyle violations.

whneo97 committed:
Add support for performing undo / redo consecutively multiple times up to the size of the history of TaskLists for the session.

whneo97 committed:
Fix checkstyle violations and make corrections for javadoc.

whneo97 committed:
Merge pull request #7 from whneo97/branch-optimise-version-history

Add support for performing undo / redo consecutively multiple times up to the size of the history of TaskLists for the session.
Fix checkstyle violations and make corrections for javadoc.

whneo97 committed:
Add About command for info on developer and Duke program.
Display red variation of duke image if an exception is thrown.
Fix bug in start and end time of random Task generator method.

whneo97 committed:
Merge branch 'branch-about' into master branch.

whneo97 committed:
Fix typographical error in undo / redo which caused major functionality loss in undo / redo commands.

whneo97 committed:
Add support for clearing cache and removing all tasklists stored in the timeline, except for the current tasklist.

whneo97 committed:
Merge branch 'branch-clear-cache' to master branch.

whneo97 committed:
Fix styling errors and javadoc parameters which have previously been missed out.

whneo97 committed:
Add software requirements and credits to about output.
Amend version in GUI title.

whneo97 committed:
Update, under information in help command, command descriptions for help, undo and redo.

whneo97 committed:
Draft user guide.

whneo97 committed:
Set theme jekyll-theme-slate

whneo97 committed:
Update User Guide

whneo97 committed:
Further updates to User Guide

whneo97 committed:
Merge branch 'master' of https://github.com/whneo97/duke

whneo97 committed:
Specify keywords and detailed functionalities in UserGuide where required

whneo97 committed:
Remove screenshot of Duke program with naming in uppercase and test print statement in MainWindow.

whneo97 committed:
Add correctly-named screenshot of program Ui with minor cropping of window to exact size.

whneo97 committed:
Update tasklists in user guide and screenshots to include only generic tasks.

whneo97 committed:
Update timings for "annual concert" examples in user guide to those shown in Ui screenshot.

whneo97 committed:
Update descriptions in help command with minor additions and include user guide website.
Update user guide example to reflect minor changes in help command.

whneo97 committed:
Verify cross-compatibility with Linux and update user guide.

whneo97 committed:
Update logo to be represented using png image to allow for cross-compatibility.
Set Calibri font also in CSS stylesheet as a safety net for fonts (to minimise possibility of fonts not loading properly on different os).
Correct mislabeled assertions in DialogBox.java.

whneo97 committed:
Merge branch 'branch-logo-png' into master.

whneo97 committed:
Update user guide with details on slight differences in font family appearances across different platforms.

whneo97 committed:
Update Ui screenshot, user guide, version and about informationto reflect changes in V2.4.
Reformat order of imports. (java imports before Duke imports)

whneo97 committed:
Fix checkstyle errors.

whneo97 committed:
Fix font compatibility issues across platforms.
Credits to Ong Bing Jue (https://github.com/bjhoohaha) for helping to spot and resolve bug causing these issues.

whneo97 committed:
Remove extra "System.out.println" in code.

whneo97 committed:
Merge branch 'branch-font-fix' to master.

whneo97 committed:
Update credits in user guide and include relevant references used.

whneo97 committed:
Update display of links in user guide.

whneo97 committed:
Minor update in user guide to maintain consistency in short command descriptions.

whneo97 committed:
Minor edit to standardise format of short command descriptions in user guide.

whneo97 committed:
Remove background at top left and right corners of Ui screenshot.

whneo97 committed:
Set background of Ui (top left and right corners) to be white.

whneo97 committed:
Revert "Set background of Ui (top left and right corners) to be white."

This reverts commit 2365534e295ee788bc20f9fc3f260199c973a146.

whneo97 committed:
Include icon for the title bar of MainWindow.

whneo97 committed:
Include duplicate detection feature in User Guide.
